crypto/rand/randfile.c

```diff
@@ -83,26 +83,37 @@ int RAND_load_file(const char *file, long bytes)
         return 0;
 
 #ifndef OPENSSL_NO_POSIX_IO
-    if (stat(file, &sb) < 0 || !S_ISREG(sb.st_mode)) {
-        RANDerr(RAND_F_RAND_LOAD_FILE, RAND_R_NOT_A_REGULAR_FILE);
-        ERR_add_error_data(2, "Filename=", file);
-        return -1;
-    }
+    /*
+     * If not a regular file, assume it's a device and don't try to
+     * read everything.
+     */
+    if (stat(file, &sb) >= 0 && !S_ISREG(sb.st_mode) && bytes < 0)
+        bytes = 256 / 8;
 #endif
+
     if ((in = openssl_fopen(file, "rb")) == NULL) {
         RANDerr(RAND_F_RAND_LOAD_FILE, RAND_R_CANNOT_OPEN_FILE);
         ERR_add_error_data(2, "Filename=", file);
         return -1;
     }
 
+    /* Read the file, in chunks. */
     for ( ; ; ) {
         if (bytes > 0)
             n = (bytes < RAND_FILE_SIZE) ? (int)bytes : RAND_FILE_SIZE;
         else
             n = RAND_FILE_SIZE;
+
         i = fread(buf, 1, n, in);
```

Comment:
I think there is still a concern about reading BUFSIZ here.  From a file, it won't matter but from a device it could.

Reply:
How about `setbuffer(in, NULL)` if it is a device?

Reply:
Is setbuffer in posix/ansi?  I can do the old clasic ```setbuf(in,null)``` if you want.

Reply:
Reading 32 bytes won't help -- fread buffers & will try to read BUF_SIZ.
setbuffer is C89 and was in K&R too.  It is used elsewhere in OpenSSL.

Reply:
From the manual on Linux:

> ```
> CONFORMING TO
>        The setbuf() and setvbuf() functions conform to C89 and C99.
> ```

I would use the shorter form...  something we already do, so...:

```
: ; git grep setbuf
CHANGES:        Use setbuf() and remove OPENSSL_NO_SETVBUF_IONBF
apps/apps.c: * On VMS, setbuf() will only take 32-bit pointers, and a compilation
apps/apps.c:    setbuf(fp, NULL);
crypto/rand/rand_egd.c:    setbuf(fp, NULL);
crypto/rand/rand_unix.c:            setbuf(fp, NULL);
```